### PR TITLE
CU-2hchqdh - Implement capped funding rates

### PR DESCRIPTION
**Pull request by 0xangelo**
## Issue
[CU-2hchqdh](https://app.clickup.com/t/2hchqdh)

## Description
Includes:
- change from paying funding from the Insurance Fund to paying from the Fee Pool
- reimplementation of `FixedPointNumber` math
  -  division that floors the quotient instead of rounding it
  - computing the remainder in a division to properly account for all funds
- incorporation of unrealized funding to margin ratio calculations
- refactorings and quality of life improvements

### CODE REVIEW

**Comment by ComposableFiB0t** [comment]
Task linked: [CU-2hchqdh Cap funding rate based on Fee Pool depth](https://app.clickup.com/t/2hchqdh)

File: frame/clearing-house/src/tests/instruments.rs [hunk 1]
```diff
@@ -160,28 +153,55 @@ proptest! {
...
+			//                 0 |                 - |  0
+			let direction = position.direction();
+			if cum_funding_delta.is_zero() || direction.is_none() {
+				assert!(unrealized_funding.is_zero());
+			} else  {
+				assert!(match direction.unwrap() {
+					Direction::Long =>
+						cum_funding_delta.is_positive() != unrealized_funding.is_positive(),
+					_ => cum_funding_delta.is_positive() == unrealized_funding.is_positive(),
+				});
```
Comment (Cardosaum) [inline frame/clearing-house/src/tests/instruments.rs:203]: What do you think about creating one test function for each case you are considering here?
The total number of random tests generated for this function would be split in 4 different non-related cases, but if you use one test function for each case you would be assessing more thoroughly each possibility.

File: frame/clearing-house/src/tests/update_funding.rs [hunk 2]
```diff
@@ -35,6 +44,34 @@ prop_compose! {
...
+	) -> (Balance, Balance) {
+		// alice_position = bob_position + net_position
+		// fees = (alice_position + bob_position) * 1%
+		//      = 2 * bob_position * 1% + net_position * 1%
+		// bob_funding = (alice_position - net_position) * 5%
+		// alice_funding = alice_position * 5%
+		//               > 2 * (alice_position - net_position) * 1% + net_position * 1%
+		//                 + (alice_position - net_position) * 5%
+		//               = alice_position * 7% - net_position * 6%
+		// <=> net_position > alice_position / 3
```
Comment (Cardosaum) [inline frame/clearing-house/src/tests/update_funding.rs:70]: Maybe add a `TODO(0xangelo): ...` here to take care of this commented section?

File: frame/clearing-house/src/types.rs [hunk 3]
```diff
@@ -69,12 +70,18 @@ pub struct Market<T: Config> {
...
+	/// Total position, in base asset, of all traders that are short. Must be negative. Used to
+	/// compute parameter adjustment costs and funding payments from/to the Clearing House
+	pub base_asset_amount_short: T::Decimal,
+	/// The latest cumulative funding rate for long positions in this market. Must be updated
+	/// periodically
+	pub cum_funding_rate_long: T::Decimal,
+	/// The latest cumulative funding rate for short positions in this market. Must be updated
+	/// periodically
+	pub cum_funding_rate_short: T::Decimal,
 	/// Amount, in quote asset, of fees collected from trades
```
Comment (Cardosaum) [inline frame/clearing-house/src/types.rs:85]: Nitpick but worth mentioning: 

Citing the [Rust Style Guide](https://doc.rust-lang.org/1.0.0/style/style/comments.html#sentence-structure) regarding comments: 

> All doc comments, including the summary line, should begin with a capital letter and end with a period, question mark, or exclamation point.

I think it might be good to add a ending period in all comments here.

**Review by Cardosaum — APPROVED** [review]
Your PR is really great :rocket: 

I just added some comments, but they are mostly nitpicks and details.

One thing I would like to comment tho that I think would be great to add in your PR (or maybe in a future one) is testing all the functions you added for math-related operations.

The math functionality you added is really great, but I think it would be important for us to ensure it is working correctly. I thought it could be a great opportunity to try using [Kani](https://github.com/model-checking/kani). 

I can help you add this tests for all math-related functions if you wish :)

As always, great work with the PR!

:ship: :skull:

File: frame/clearing-house/src/lib.rs [hunk 4]
```diff
@@ -941,71 +988,49 @@ pub mod pallet {
...
-			direction: Direction,
-		) -> Result<(), DispatchError> {
-			let base_amount_decimal: T::Decimal = base_asset_swapped.into_decimal()?;
-			market.net_base_asset_amount.try_add_mut(&match direction {
-				Direction::Long => base_amount_decimal,
-				Direction::Short => base_amount_decimal.neg(),
-			})?;
-			Ok(())
+				.try_mul(&market.taker_fee)?
+				.try_div(&10_000_u32.into())
```
[the remark this answers is not in the document]
Comment (kevin-composable) [reply frame/clearing-house/src/lib.rs:992]: I second this

File: frame/clearing-house/src/tests/open_position.rs [hunk 5]
```diff
@@ -663,6 +666,55 @@ proptest! {
 		});
 	}
 
+	#[test]
+	fn margin_ration_takes_unrealized_funding_into_account(direction in any_direction()) {
```
Comment (kevin-composable) [inline frame/clearing-house/src/tests/open_position.rs:670]: should this be margin_ratio_...

**Review by kevin-composable — APPROVED** [review]
Looks great. I didn't really have anything that NEEDED to be changed. Fantastic job as always Angelo

File: frame/clearing-house/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (0xangelo) [reply frame/clearing-house/src/lib.rs:992]: Good suggestion!

File: frame/clearing-house/src/math.rs [hunk 6]
```diff
@@ -54,53 +146,82 @@ where
...
+			false => Overflow,
+		};
+
+		let lhs: I129 = self.into_inner().into();
+		let rhs: I129 = other.into_inner().into();
+		let negative = lhs.negative != rhs.negative;
+
+		let accuracy = Self::DIV.unique_saturated_into();
+		let q = multiply_by_rational(lhs.value, accuracy, rhs.value).map_err(|_| error)?;
+		Ok(Self::from_inner(from_i129(I129 { value: q, negative }).ok_or(error)?))
```
[the remark this answers is not in the document]
Comment (0xangelo) [reply frame/clearing-house/src/math.rs:202]: Good catch, I'll try using a closure instead

File: frame/clearing-house/src/tests/instruments.rs (hunk 1 above)
Comment (0xangelo) [reply frame/clearing-house/src/tests/instruments.rs:203]: Yeah, that's true. It could be the case that the randomly generated inputs don't cover the edge case of `0` too. I'll split it

File: frame/clearing-house/src/tests/update_funding.rs (hunk 2 above)
Comment (0xangelo) [reply frame/clearing-house/src/tests/update_funding.rs:70]: This is actually a 'derivation' of the fact that `net_position` has to be greater than `alice_position / 3` (with these fees and funding rates) for the resulting funding payments to be capped. I should make that more clear

File: frame/clearing-house/src/types.rs (hunk 3 above)
Comment (0xangelo) [reply frame/clearing-house/src/types.rs:85]: Seems there are a lot of periods missing :grimacing:

File: frame/clearing-house/src/tests/open_position.rs (hunk 5 above)
Comment (0xangelo) [reply frame/clearing-house/src/tests/open_position.rs:670]: Oops 🤭

**Comment by 0xangelo** [comment]
> One thing I would like to comment tho that I think would be great to add in your PR (or maybe in a future one) is testing all the functions you added for math-related operations.
> 
> The math functionality you added is really great, but I think it would be important for us to ensure it is working correctly. I thought it could be a great opportunity to try using [Kani](https://github.com/model-checking/kani).

Great suggestion. I added it to [ClickUp](https://app.clickup.com/t/2j97q70)

File: frame/clearing-house/src/tests/update_funding.rs (hunk 2 above)
Comment (Cardosaum) [reply frame/clearing-house/src/tests/update_funding.rs:70]: Oh I see... :+1:

File: frame/clearing-house/src/tests/update_funding.rs [hunk 7]
```diff
@@ -59,15 +59,19 @@ prop_compose! {
 		alice_position in Just(alice_position),
 		net_position in balance_in_open_interval(alice_position / 3, alice_position),
 	) -> (Balance, Balance) {
+		// Assuming a taker fee of 1% and a funding rate of 5% in favor of Alice's position, the net
+		// position should be at least one third of Alice's position for funding to be capped
+		// (for Alice). Derivation:
```
Comment (Cardosaum) [inline frame/clearing-house/src/tests/update_funding.rs:64]: Now I get it :sweat_smile: 

Thanks! :)

**Review by Cardosaum — APPROVED** [review]

File: frame/clearing-house/src/tests/instruments.rs (hunk 1 above)
Comment (0xangelo) [reply frame/clearing-house/src/tests/instruments.rs:203]: Actually, I manually set the specific cases in the input strategies:
```rust
		(base_amount, quote_amount) in prop_oneof![
			Just((-10, -100)),
			Just((0, 0)),
			Just((10, 100))
		],
		cum_funding_delta in
			prop_oneof![Just(-1), Just(0), Just(1)].prop_map(|n| FixedI128::from((n, 10))),
```
So there's no real benefit of splitting it now
